Plugins: Angular detector: Add database cache store for remote patterns

xnyo opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

7. If your PR content should be added to the What's New document for the next major or minor release, add the **add to what's new** label to your PR. Note that you should add this label to the main PR that introduces the feature; do not add this label to smaller PRs for the feature.

-->

**What is this feature?**

2nd part of #69843 .

Adds caching for angular detection patterns fetched from GCOM, similar to plugins signing keys.

It uses a kvstore to save the patterns in the database, then they are converted to angular detectors in-memory to be used with the angular detector.

There's also a background service that periodically fetches and stores new patterns from GCOM into the database, and updates the in-memory detectors accordingly.

All the GCOM related code is in pluginsintegration.

The GCOM fetcher is still under a feature toggle, and it (along with the background service) is disabled if the feature flag is not present.

**Other changes:**

- Some files from the previous PR were moved/renamed.
- Reduces the context timeout for angular detection from 1 minute to 10 seconds.

**Config canges:**

Enable the feature toggle if you want to test locally:

```ini
[feature_toggles]
enable = pluginsDynamicAngularDetectionPatterns
```

**Why do we need this feature?**

Use one network request rather than one per plugin, making startup faster.

**Who is this feature for?**

Grafana users and operators relying on Angular plugin

**Which issue(s) does this PR fix?**:

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Related to #68974 

**Special notes for your reviewer:**

Please check that:
- [x] It works as expected from a user's perspective.
- [x] If this is a pre-GA feature, it is behind a feature toggle.
- [x] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

xnyo committed:
Plugins: Angular detector: Remote patterns fetching

xnyo committed:
Renamed PatternType to GCOMPatternType

xnyo committed:
Renamed files

xnyo committed:
Renamed more files

xnyo committed:
Moved files again

xnyo committed:
Add type checks, unexport GCOM structs

xnyo committed:
Cache failures, update log messages, fix GCOM URL

xnyo committed:
Fail silently for unknown pattern types, update docstrings

xnyo committed:
Fix tests

xnyo committed:
Rename gcomPattern.Value to gcomPattern.Pattern

xnyo committed:
Refactoring

xnyo committed:
Add FlagPluginsRemoteAngularDetectionPatterns feature flag

xnyo committed:
Merge branch 'main' into giuseppe/angular-deprecation/remote-patterns

xnyo committed:
Fix tests

xnyo committed:
Merge branch 'main' into giuseppe/angular-deprecation/remote-patterns

xnyo committed:
Merge branch 'main' into giuseppe/angular-deprecation/remote-patterns

xnyo committed:
Re-generate feature flags

xnyo committed:
Add TestProvideInspector, renamed TestDefaultStaticDetectorsInspector

xnyo committed:
Add TestProvideInspector

xnyo committed:
Add TestContainsBytesDetector and TestRegexDetector

xnyo committed:
Renamed getter to provider

xnyo committed:
More tests

xnyo committed:
TestStaticDetectorsProvider, TestSequenceDetectorsProvider

xnyo committed:
GCOM tests

xnyo committed:
Lint

xnyo committed:
Merge branch 'main' into giuseppe/angular-deprecation/remote-patterns

xnyo committed:
Made detector.detect unexported, updated docstrings

xnyo committed:
Allow changing grafana.com URL

xnyo committed:
Merge branch 'main' into giuseppe/angular-deprecation/remote-patterns

xnyo committed:
Fix API path, add more logs

xnyo committed:
Update tryUpdateRemoteDetectors docstring

xnyo committed:
Use angulardetector http client

xnyo committed:
Return false, nil if module.js does not exist

xnyo committed:
Chore: Split angualrdetector into angularinspector and angulardetector packages

Moved files around, changed references and fixed tests:
- Split the old angulardetector package into angular/angulardetector and angular/angularinspector
- angulardetector provides the detection structs/interfaces (Detector, DetectorsProvider...)
- angularinspector provides the actual angular detection service used directly in pluginsintegration
- Exported most of the stuff that was private and now put into angulardetector, as it is not required by angularinspector

xnyo committed:
Renamed detector.go -> angulardetector.go and inspector.go -> angularinspector.go

Forgot to rename those two files to match the package's names

xnyo committed:
Renamed angularinspector.ProvideInspector to angularinspector.ProvideService

xnyo committed:
Renamed "harcoded" to "static" and "remote" to "dynamic"

from PR review, matches the same naming schema used for signing keys fetching

xnyo committed:
WIP: Angular: cache patterns in db, moved gcom into pluginsintegration

More similar to signing keys fetching

xnyo committed:
Rename package, refactoring

xnyo committed:
Merge branch 'main' into giuseppe/angular-deprecation/remote-patterns-db

xnyo committed:
try to solve circular import

xnyo committed:
Merge branch 'main' into giuseppe/angular-deprecation/remote-patterns

xnyo committed:
Fix merge conflict on updated angular patterns

xnyo committed:
Fix circular imports

xnyo committed:
Fix wire gen

xnyo committed:
Add docstrings, refactoring

xnyo committed:
Removed angualrdetectorsprovider dependency into angularpatternsstore

xnyo committed:
Moved GCOM test files

xnyo committed:
Removed GCOM cache

xnyo committed:
Renamed Detect to DetectAngular and Detector to AngularDetector

xnyo committed:
Merge remote-tracking branch 'origin/giuseppe/angular-deprecation/remote-patterns' into giuseppe/angular-deprecation/remote-patterns

xnyo committed:
Fix call to NewGCOMDetectorsProvider in newDynamicInspector

xnyo committed:
Removed unused test function newError500GCOMScenario

xnyo committed:
Added angularinspector service definition in pluginsintegration

xnyo committed:
Merge branch 'giuseppe/angular-deprecation/remote-patterns' into giuseppe/angular-deprecation/remote-patterns-db

xnyo committed:
refactoring

xnyo committed:
Merge branch 'main' into giuseppe/angular-deprecation/remote-patterns-db

xnyo committed:
lint

xnyo committed:
Fix angularinspector TestProvideService

xnyo committed:
cleanup

xnyo committed:
Await initial restore

xnyo committed:
Register dynamicAngularDetector background service

xnyo committed:
Removed static detectors provider from pluginsintegration

xnyo committed:
Add tests for kvstore

xnyo committed:
Add more tests

xnyo committed:
order imports in dynamic_test.go

xnyo committed:
Fix potential panic in dynamic_test

xnyo committed:
Add "runs the job periodically" test

xnyo committed:
lint

xnyo committed:
add timeout to test

xnyo committed:
refactoring

xnyo committed:
Removed context.Context from DetectorsProvider

xnyo committed:
Refactoring, ensure angular dynamic background service is not started if feature flag is off

xnyo committed:
Fix deadlock on startup

xnyo committed:
Fix angulardetectorsprovider tests

xnyo committed:
Merge branch 'main' into giuseppe/angular-deprecation/remote-patterns-db

xnyo committed:
Revert "Removed context.Context from DetectorsProvider"

This reverts commit 4e8c6dded70844709400fa0ce4ce45e66c8458ca.

xnyo committed:
Fix wrong argument number in dynamic_teset

xnyo committed:
Standardize gcom http client

xnyo committed:
Reduce context timeout for angular inspector in plugins loader

xnyo committed:
Simplify initial restore logic

xnyo committed:
Fix dynamic detectors provider tests

xnyo committed:
Chore: removed angulardetector/provider.go

xnyo committed:
Add more tests

xnyo commented on pkg/services/pluginsintegration/angulardetectorsprovider/dynamic.go:
```diff
@@ -0,0 +1,302 @@
+}
+
+// Same configuration as pkg/plugins/repo/client.go
+func makeHttpClient() http.Client {
```
Duplicated code from the signing keys fetcher. This should probably be shared with the signing keys service but I'll leave it for a follow-up PR.

xnyo replied:
:+1: yep that should work! I'll try that, thanks

xnyo replied:
It is also used to save the fetched value in the store afterwards, so I think it makes sense to keep it, but maybe rename it:

```go
gcomFetchTimeout    = time.Minute * 1
```

:point_down: 

```go
backgroundJobTimeout    = time.Minute * 1
```

WDYT?

andresmgot replied:
> It is also used to save the fetched value in the store afterwards

Same here about the timeout, I don't think we need to worry about that. Only about doing http requests over the internet.

andresmgot commented on pkg/services/pluginsintegration/angularpatternsstore/store.go:
```diff
@@ -0,0 +1,85 @@
+	if err := s.kv.Set(ctx, keyPatterns, string(b)); err != nil {
+		return fmt.Errorf("kv set: %w", err)
+	}
+	if err := s.kv.Set(ctx, keyLastUpdated, time.Now().Format(time.RFC3339)); err != nil {
+		return fmt.Errorf("kv last updated set: %w", err)
+	}
```
nice idea to wrap `Set` with the update of `LastUpdated`. We can do that for the `Set` and `Delete` methods of the generic `store`

andresmgot reviewed (commented):
Nice job! I think we can simplify a bit things here, let me know what you think.

xnyo commented on pkg/services/pluginsintegration/angulardetectorsprovider/dynamic.go:
```diff
@@ -0,0 +1,302 @@
+	// Use setDetectors and ProvideDetectors to write/read this value.
+	detectors []angulardetector.AngularDetector
+
+	// mux is the mutex used to read/write the cached detectors in a concurrency-safe way.
+	mux sync.RWMutex
+
+	// initialRestore is a channel that can be used to await the initial restore to be done.
+	// It is closed when the initial restore from cache is completed.
+	// To wait for the initial restore to be done, simply read a value from this channel.
+	initialRestore chan struct{}
```
It could be done with the `mux`, but it would require locking the `mux` before starting the [goroutine responsible for the initial restore](https://github.com/grafana/grafana/pull/70693/files/080114d63ce67f5c49c6010c9270b2ce1a383dd5#diff-a24d5d308965e3b364ed7a21123ba8917b5f644d9940230e9de564bb82e63f76R83) in `ProvideDynamic`, because there is no guarantee that the goroutine doing the initial restore runs before `ProvideDetectors` (where we await the initial restore) is called.

If `ProvideDetectors` runs first, the `mux` is unlocked and it will return `nil`, which we don't want

andresmgot replied:
> it would require locking the `mux` before starting the [goroutine responsible for the initial restore](https://github.com/grafana/grafana/pull/70693/files/080114d63ce67f5c49c6010c9270b2ce1a383dd5#diff-a24d5d308965e3b364ed7a21123ba8917b5f644d9940230e9de564bb82e63f76R83)

That sounds fine to me, any reason not to?

xnyo replied:
I'm not a big fan of locking a mutex in one goroutine and unlocking in a different one, but it makes things cleaner I guess. I have changed it to use the mutex and removed the channel :+1:

xnyo commented on pkg/services/pluginsintegration/angulardetectorsprovider/dynamic.go:
```diff
@@ -0,0 +1,302 @@
+	// To wait for the initial restore to be done, simply read a value from this channel.
+	initialRestore chan struct{}
+
+	// backgroundJob is the implementation of the background job. This is periodically invoked by Run().
+	backgroundJob backgroundJob
```
Yes, it's for tests and I'm not a big fan of this either. However, the PR went to different iterations and with the current one I think it can be achieved like you suggested (as long as we lock the mutex before fetching patterns from GCOM and release it after we update the store etc). I will take a look :+1:

xnyo commented on pkg/services/pluginsintegration/angulardetectorsprovider/dynamic.go:
```diff
@@ -0,0 +1,302 @@
+	// backgroundJob is the implementation of the background job. This is periodically invoked by Run().
+	backgroundJob backgroundJob
+
+	// backgroundJobInterval is the interval between the periodic background job calls.
+	backgroundJobInterval time.Duration
```
I think it's used in tests to change it to a shorter one, to test that the background job calls the function periodically. We don't really have a "cron"-like background job in Grafana backend as far as I know, so I wanted to test that the background service periodically calls the update function

andresmgot replied:
I see, just a matter of style but I would rename `defaultBackgroundJobInterval` to just `backgroundJobInterval` and make it a variable (not a const). Then you can override that value in tests. Also adding a comment saying that it's defined as variable for the tests.

xnyo replied:
sounds good, I have removed `backgroundJobInterval` and used a variable in tests instead

xnyo commented on pkg/services/pluginsintegration/angulardetectorsprovider/dynamic.go:
```diff
@@ -0,0 +1,302 @@
+	// Perform the initial restore from db without blocking
+	go func() {
+		d.log.Debug("Restoring cache")
+		ctx, canc := context.WithTimeout(context.Background(), cacheRestoreTimeout)
+		defer canc()
```
Are you suggesting to inline the `cacheRestoreTimeout` const or remove the timeout entirely?

This unfortunately requires a timeout because we do not have a base `ctx` passed to `ProvideDynamic`. If we don't have a timeout, and for some reason the store retrieval times out, I think Grafana will deadlock on startup.

I'm not sure if the store already applies some kind of timeout tho, maybe you have more *context* on this (no pun intended :stuck_out_tongue: )

andresmgot replied:
yes, I mean removing it entirely. As far as I can tell this is just reading from the DB, so it should be fast (nothing over the internet). In any case, I don't think we need to worry if the database is not available, plenty of other things would fail in that case :)

xnyo commented on pkg/services/pluginsintegration/angulardetectorsprovider/dynamic.go:
```diff
@@ -0,0 +1,302 @@
+	d.mux.Unlock()
+}
+
+// tryUpdateDetectors will attempt to fetch the patterns from GCOM, convert them to detectors,
+// store the patterns in the database and update the cached detectors.
+func (d *Dynamic) tryUpdateDetectors(ctx context.Context) {
```
:+1: I will try that!

xnyo commented on pkg/services/pluginsintegration/angulardetectorsprovider/dynamic_updater.go:
```diff
@@ -0,0 +1,31 @@
+	registry.BackgroundService
+}
+
+// noOpDynamicUpdater is a DynamicUpdater that immediately returns nil.
+type noOpDynamicUpdater struct{}
+
+// Run returns nil.
+func (s *noOpDynamicUpdater) Run(_ context.Context) error {
+	return nil
+}
```
:+1: will do!

xnyo committed:
Removed backgroundJob interface, PR review feedback

xnyo committed:
Update tests

xnyo committed:
PR review feedback: remove ErrNoCachedValue from kv store Get

xnyo committed:
Update tests

xnyo committed:
PR review feedback: add IsDisabled and remove nop background srevice

xnyo committed:
Update tests

andresmgot commented on pkg/services/pluginsintegration/angulardetectorsprovider/dynamic_test.go:
```diff
@@ -0,0 +1,473 @@
+package angulardetectorsprovider
```
up to you but this file is not even rendered in GitHub because it's too large :) (so I am not reviewing it line by line)  

I see that you are testing internal functions though, that's generically not desired because it makes the code more difficult to update.

Next time try to use only the exposed functions, it's easier to do table testing if only those are used in tests.

marefr replied:
For unit tests, testing internal functions could be fine and how we do unit tests around the code base. However, feels like part of these tests are almost like integration tests (the background job processing for example) so could perhaps be split up/more clearly separated by for example using multiple `Test*` functions?

xnyo replied:
sounds good, I have moved the background services tests in a separate test function as suggested

andresmgot commented on pkg/services/pluginsintegration/angularpatternsstore/store.go:
I'm actually not suggesting a follow-up but the other way around (a pre-PR?). That would make this PR easier to review (less code). In any case I don't want to block this PR because of that but you can work in parallel, if the other PR is ready before this one you can apply the changes here, if not it's okay to remove this code later.

marefr commented on pkg/services/pluginsintegration/angulardetectorsprovider/dynamic.go:
```diff
@@ -0,0 +1,282 @@
+		return fmt.Errorf("get cached value: %w", err)
+	}
+	// Try to unmarshal, convert to detectors and set local cache
+	if err := json.Unmarshal([]byte(rawCached), &cachedPatterns); err != nil {
```
Should this logic perhaps live in the store to provide a nicer API and make the store operate on and return strongly typed types rather than `any`?

marefr reviewed (commented):
Great work. I agree with Andres that some things can probably be simplified

xnyo committed:
Merge branch 'main' into giuseppe/angular-deprecation/remote-patterns-db

xnyo committed:
Remove initialRestore channel, use mux instead

xnyo committed:
Removed backgroundJobInterval, use package-level variable instead

xnyo committed:
Add TestDynamicAngularDetectorsProviderBackgroundService

xnyo commented on pkg/plugins/manager/loader/loader.go:
```diff
@@ -191,12 +191,12 @@ func (l *Loader) loadPlugins(ctx context.Context, src plugins.PluginSource, foun
 			var err error
 
-			cctx, canc := context.WithTimeout(ctx, time.Minute*1)
+			cctx, canc := context.WithTimeout(ctx, time.Second*10)
```
We forbid loading Angular plugins if they are disabled (which is not the default behaviour by default for now), so unfortunately we cannot run the detections in the background. See #69679

andresmgot commented on pkg/services/pluginsintegration/angulardetectorsprovider/dynamic.go:
```diff
@@ -63,19 +58,18 @@ func ProvideDynamic(cfg *config.Cfg, store angularpatternsstore.Service, feature
 	}
 
 	// Perform the initial restore from db without blocking
+	d.mux.Lock()
```
what if you don't `Lock` here but in `setDetectorsFromCache`? I am thinking on something like:

```patch
func (d *Dynamic) setDetectorsFromCache(ctx context.Context) error {
	var cachedPatterns GCOMPatterns
+ d.mux.RLock()
	rawCached, ok, err := d.store.Get(ctx)
+ d.mux.RUnlock()
	if !ok {
		// No cached value found, do not alter in-memory detectors
		return nil
	}
	if err != nil {
		return fmt.Errorf("get cached value: %w", err)
	}
	// Try to unmarshal, convert to detectors and set local cache
	if err := json.Unmarshal([]byte(rawCached), &cachedPatterns); err != nil {
		return fmt.Errorf("json unmarshal: %w", err)
	}
	cachedDetectors, err := d.patternsToDetectors(cachedPatterns)
	if err != nil {
		return fmt.Errorf("convert to detectors: %w", err)
	}
-	// Do not use setDetectors or it will cause a deadlock
- 	d.detectors = cachedDetectors
+ d.setDetectors()
	return nil
}
```

xnyo replied:
Unfortunately this doesn't work. The first call of ProvideDetectors must await the initial restore, that's why I had the channel in the first place.

If we want to use a mutex, we have to lock it before launching the [goroutine that fetches the values from the database in the background](https://github.com/grafana/grafana/blob/giuseppe/angular-deprecation/remote-patterns-db/pkg/services/pluginsintegration/angulardetectorsprovider/dynamic.go#L58), otherwise whoever calls ProvideDetectors may find it unlocked (since there's no guarantee the "restore" goroutine will run first) and return an empty value (which is wrong).

I also have [a test](https://github.com/grafana/grafana/blob/giuseppe/angular-deprecation/remote-patterns-db/pkg/services/pluginsintegration/angulardetectorsprovider/dynamic_test.go#L67) for this, with your suggested changes the test fails.

Alternatively, we can perform the restore in the main goroutine, but this will increase the Grafana startup times significantly if GCOM is slow to respond:

```diff
--- a/pkg/services/pluginsintegration/angulardetectorsprovider/dynamic.go
+++ b/pkg/services/pluginsintegration/angulardetectorsprovider/dynamic.go
@@ -52,17 +52,11 @@ func ProvideDynamic(cfg *config.Cfg, store angularpatternsstore.Service, feature
                httpClient: makeHttpClient(),
                baseURL:    cfg.GrafanaComURL,
        }
-
-       // Perform the initial restore from db without blocking
-       d.mux.Lock()
-       go func() {
-               d.log.Debug("Restoring cache")
-               if err := d.setDetectorsFromCache(context.Background()); err != nil {
-                       d.log.Warn("Cache restore failed", "error", err)
-               }
-               d.mux.Unlock()
-       }()
-
+       // Perform the initial restore from db, return the service once restore is done
+       d.log.Debug("Restoring cache")
+       if err := d.setDetectorsFromCache(context.Background()); err != nil {
+               d.log.Warn("Cache restore failed", "error", err)
+       }
        return d, nil
 }
 
@@ -175,8 +169,7 @@ func (d *Dynamic) setDetectorsFromCache(ctx context.Context) error {
        if err != nil {
                return fmt.Errorf("convert to detectors: %w", err)
        }
-       // Do not use setDetectors or it will cause a deadlock
-       d.detectors = cachedDetectors
+       d.setDetectors(cachedDetectors)
        return nil
 }
```

I would either keep it like it is currently (locking before launching the other goroutine), or revert to using a channel if it's more confusing (I personally think using a channel is more idiomatic, but no strong opinions on this)

andresmgot replied:
> Alternatively, we can perform the restore in the main goroutine, but this will increase the Grafana startup times significantly if GCOM is slow to respond

I may not be seeing it but there is no GCOM request in `setDetectorsFromCache`? Only getting from the `store` which should be fast?

xnyo replied:
Sorry I meant database not GCOM. If we can assume that the database is always fast, then sure, I will change it

andresmgot replied:
:+1: yes, I don't think we need to worry about that

xnyo committed:
Removed timeouts

xnyo committed:
pr review feedback: restore from store before returning the service

xnyo committed:
Update tests

xnyo commented:
I have addressed the feedback and moved the initialization before returning the service and removed the timeouts on the contexts. I will leave the generic store refactoring (and the other store changes suggested by Marcus) for another PR to get this one shipped if that's ok.

andresmgot commented on pkg/services/pluginsintegration/angulardetectorsprovider/dynamic.go:
```diff
@@ -0,0 +1,253 @@
+	if err := d.setDetectorsFromCache(context.Background()); err != nil {
+		d.log.Warn("Cache restore failed", "error", err)
+	}
+	return d, nil
```
if you add a log line here (in case of success) you can easily get the time spent here when running this on cloud

xnyo replied:
Good idea! Will add :+1:

andresmgot replied:
the log line already includes the timestamp so there is no need to add the duration :shrug:

xnyo replied:
I have seen it logged like this in a few places for example:

- https://github.com/grafana/grafana/blob/giuseppe/angular-deprecation/remote-patterns-db/pkg/services/updatechecker/plugins.go#L111

- https://github.com/grafana/grafana/blob/giuseppe/angular-deprecation/remote-patterns-db/pkg/infra/usagestats/service/usage_stats.go#L124

- https://github.com/grafana/grafana/blob/giuseppe/angular-deprecation/remote-patterns-db/pkg/services/cleanup/cleanup.go#L120
- https://github.com/grafana/grafana/blob/giuseppe/angular-deprecation/remote-patterns-db/pkg/services/ngalert/backtesting/engine.go#L118

I guess it's good for consistency?

andresmgot replied:
yep, no problem

andresmgot reviewed (approved):
LGTM, let's see the follow-up PR

xnyo committed:
Log duration on startup restore and cron run

xnyo committed:
Switch cron job start log to debug level

xnyo committed:
Do not attempt to restore if disabled

xnyo committed:
Merge branch 'main' into giuseppe/angular-deprecation/remote-patterns-db
